### PR TITLE
[SPARK-23687][SS] Add a memory source for continuous processing.

### PR DESCRIPTION
**Pull request by jose-torres**
## What changes were proposed in this pull request?

Add a memory source for continuous processing.

Note that only one of the ContinuousSuite tests is migrated to minimize the diff here. I'll submit a second PR for SPARK-23688 to change the rest and get rid of waitForRateSourceTriggers.

## How was this patch tested?

unit test

### CODE REVIEW

**Comment by jose-torres** [comment]
@tdas @zsxwing

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+ *  * ContinuousMemoryStream maintains a list of records for each partition. addData() will
+ *    distribute records evenly-ish across partitions.
+ *  * ContinuousMemoryStreamRecordBuffer is set up as an endpoint for partition-level
+ *    ContinuousMemoryStreamDataReader instances to poll. It returns the record at the specified
+ *    offset within the list, or null if that offset doesn't yet have a record.
+ */
+
+private case class GetRecord(offset: ContinuousMemoryStreamPartitionOffset)
+
+private class ContinuousMemoryStreamRecordBuffer[A](
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: This is not really a buffer as it only refers to buffers that are externally managed. Its only serves as an endpoint to fetch stuff from the buffer. I would prefer one way or the other - either all the buffer management and endpoint management should be inside the class (adding and fetching, synchronized by this class) OR this is an inner class of ContinuousMemoryStream whose only purpose is to be endpoint relaying fetch requests.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+  override def mergeOffsets(offsets: Array[PartitionOffset]): ContinuousMemoryStreamOffset = {
+    ContinuousMemoryStreamOffset {
+      offsets.map {
+        case ContinuousMemoryStreamPartitionOffset(part, num) => (part, num)
+      }.toMap
+    }
+  }
+
+  var endpointRef: RpcEndpointRef = _
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: private.
and volatile to avoid issues if stop is called from a different thread.

Comment (tdas) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: also its good programming practice to consolidate all the fields to a single location.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+  def createContinuousReader(
+      schema: Optional[StructType],
+      checkpointLocation: String,
+      options: DataSourceOptions): ContinuousReader = {
+    this
+  }
+
+  // MemoryStreamBase implementation
+
+  protected val logicalPlan =
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: nit: same comment as above, consolidate vals, unless there is a really good reason.

**Review by tdas — CHANGES_REQUESTED** [review]
Overall direction looks good. My major comment is the class structure. Ideally, I would like to see all the MemoryStream related classes (MicroBatch and Continuous) named consistently (all starts with same prefix like MemoryStream), and located in the same place (execution/streaming/sources) so that its easy to find and associate (similar to Kafka classes right now). 
However, that requires requires renaming existing class like MemoryStream to MemoryStreamMicroBatchReader which may require changing it in many usage places. That's not good. And intermediate compromise may be one of the following
1. Rename the main case class MemoryStream to MemoryStreamMicroBatchReader, but keep the object MemoryStream. I think most usages refer to the object. 
2. Rename MemoryStream class and object to MemoryStreamMicroBatchReader, and add a package level alias MemoryStream (similar to how DataFrame is alias of Dataset[Row]).

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala [hunk 4]
```diff
@@ -53,32 +53,24 @@ class ContinuousSuiteBase extends StreamTest {
...
 
-  test("basic rate source") {
-    val df = spark.readStream
-      .format("rate")
-      .option("numPartitions", "5")
-      .option("rowsPerSecond", "5")
-      .load()
-      .select('value)
+  test("basic") {
+    val input = MemoryStream[Int]
```
[the remark this answers is not in the document]
Comment (jose-torres) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: What I eventually want to be able to do is to run all our existing test suites in continuous mode, to ensure that continuous processing can do all the same things as microbatch execution.

If you're not convinced, we should take this change out of the PR and punt on it. But I think we will eventually need something much like this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+import org.apache.spark.sql.sources.v2.reader.streaming.{ContinuousDataReader, ContinuousReader, Offset, PartitionOffset}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.util.RpcUtils
+
+/**
+ * The overall strategy here is:
+ *  * ContinuousMemoryStream maintains a list of records for each partition. addData() will
+ *    distribute records evenly-ish across partitions.
+ *  * ContinuousMemoryStreamRecordBuffer is set up as an endpoint for partition-level
+ *    ContinuousMemoryStreamDataReader instances to poll. It returns the record at the specified
```
[the remark this answers is not in the document]
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: To poll from "partition-level ContinuousMemoryStreamDataReader instances". I can say executor-side instead.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala [hunk 6]
```diff
@@ -0,0 +1,212 @@
+/*
```
[the remark this answers is not in the document]
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: I think the consistent naming makes sense for the sources where they're automatically injected in planning, since they both have the same .format() option and thus the same name in some sense.

For memory streams it's different; the user explicitly instantiates the reader. So I think that making the names look similar is confusing, and I don't think that "MemoryStreamMicroBatchReader" is a good name for something explicitly instantiated.

**Comment by jose-torres** [comment]
updated

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (tdas) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: I think that's too much to hope for right now. We can do that later. For now, let's make everything explicitly ContinuousMemoryStream.

[the remark this answers is not in the document]
Comment (jose-torres) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: This has been done as of the current commit. Some uses of MemoryStreamBase are still needed in order to make the AddData test handle work.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+  }
+
+  override def setStartOffset(start: Optional[Offset]): Unit = synchronized {
+    // Inferred initial offset is position 0 in each partition.
+    startOffset = start.orElse {
+      ContinuousMemoryStreamOffset((0 until NUM_PARTITIONS).map(i => (i, 0)).toMap)
+    }.asInstanceOf[ContinuousMemoryStreamOffset]
+  }
+
+  override def getStartOffset: Offset = startOffset
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: If set `setStartOffset` is synchronized then `getStartOffset` should also be synchronized.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+    StreamingRelationV2(this, "memory", Map(), attributes, None)(sqlContext.sparkSession)
+
+  // ContinuousReader implementation
+
+  @GuardedBy("this")
+  private val records = Seq.fill(NUM_PARTITIONS)(new ListBuffer[A])
+
+  private val recordBuffer = new ContinuousMemoryStreamRecordBuffer()
+
+  private var startOffset: ContinuousMemoryStreamOffset = _
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: This probably should be `GuardedBy("this")` as well.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+  }
+
+  override def getStartOffset: Offset = startOffset
+
+  override def deserializeOffset(json: String): ContinuousMemoryStreamOffset = {
+    ContinuousMemoryStreamOffset(Serialization.read[Map[Int, Int]](json))
+  }
+
+  override def mergeOffsets(offsets: Array[PartitionOffset]): ContinuousMemoryStreamOffset = {
+    ContinuousMemoryStreamOffset {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: super nit: `SomeObject.apply(anotherObject)` is usually called with with `(...)` not `{...}`. Only when you are passing a function literal do we use `{...}`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+  override def reset(): Unit = synchronized {
+    records.foreach(_.clear())
+    startOffset = ContinuousMemoryStreamOffset((0 until NUM_PARTITIONS).map(i => (i, 0)).toMap)
+  }
+
+  /**
+   * Endpoint for executors to poll for records.
+   */
+  private class ContinuousMemoryStreamRecordBuffer extends ThreadSafeRpcEndpoint {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: nit: I commented on this earlier, this class is not a buffer, its an endpoint. 

Also, I approve moving this inside the `ContinuousMemoryStream` class. Though in that case, the name can be shortened by removing `ContinuousMemoryStream` prefix.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+    override val rpcEnv: RpcEnv = SparkEnv.get.rpcEnv
+
+    override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
+      case GetRecord(ContinuousMemoryStreamPartitionOffset(part, index)) =>
+        ContinuousMemoryStream.this.synchronized {
+          val buf = records(part)
+
+          val record =
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: nit: Can this be fit into a single line?
`val record = if (buf.size <= index) None else Some(buf(index))` should fit.
also, extra line above.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+            }
+          context.reply(record.map(Row(_)))
+        }
+    }
+  }
+}
+
+object ContinuousMemoryStream {
+  def recordBufferName(memoryStreamId: Int): String =
+    s"ContinuousMemoryStreamRecordReceiver-$memoryStreamId"
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: nit: this term `ContinuousMemoryStreamRecordReceiver` does not seem to match with any class. neither does it have "buffername" in it.

add docs what it means.

Comment (tdas) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: its only used inside `createDataReaderFactories()`, so why even have this a separate method.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+      current = endpoint.askSync[Option[Row]](
+          GetRecord(ContinuousMemoryStreamPartitionOffset(partition, currentOffset)))
+    }
+    currentOffset += 1
+    true
+  }
+
+  override def get(): Row = current.get
+
+  override def close(): Unit = {}
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: does the `endpoint` require closing  or cleaning up?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala (hunk 10 above)
Comment (tdas) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: Actually, it is actually confusing to see `recordBuffer` at places like `recordBuffer.rpcEnv.setupEnpoint...`
I would prefer it be named as an endpoint rather than a buffer.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (tdas) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: However, I stand by the other point of all the sources being colocated in streaming/sources. You missed changing that.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+  override def mergeOffsets(offsets: Array[PartitionOffset]): ContinuousMemoryStreamOffset = {
+    ContinuousMemoryStreamOffset {
+      offsets.map {
+        case ContinuousMemoryStreamPartitionOffset(part, num) => (part, num)
+      }.toMap
+    }
+  }
+
+  override def createDataReaderFactories(): ju.List[DataReaderFactory[Row]] = {
+    synchronized {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: Why is this method synchronized? Its not accessing the buffer.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala [hunk 15]
```diff
@@ -148,7 +167,7 @@ case class MemoryStream[A : Encoder](id: Int, sqlContext: SQLContext)
     }
   }
 
-  private def generateDebugString(
+  protected def generateDebugString(
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala]: why this change?

**Review by tdas — CHANGES_REQUESTED** [review]
Its looks good overall, just a bunch of nits.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala (hunk 13 above)
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: No - IIUC it's a stateless reference to the remote endpoint.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala (hunk 14 above)
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousMemoryStream.scala]: It has to be now that startOffset is GuardedBy("this"). (I think it's correct that startOffset should have been guarded.)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala (hunk 15 above)
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala]: honestly not sure why this ended up happening - un-changed it

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala [hunk 16]
```diff
@@ -43,20 +45,51 @@ object MemoryStream {
   protected val currentBlockId = new AtomicInteger(0)
   protected val memoryStreamId = new AtomicInteger(0)
 
-  def apply[A : Encoder](implicit sqlContext: SQLContext): MemoryStream[A] =
+  def apply[A : Encoder](
+      implicit sqlContext: SQLContext,
+      trigger: Trigger = Trigger.ProcessingTime(0)): MemoryStream[A] =
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala]: You are not using this `trigger` anymore.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala [hunk 17]
```diff
@@ -53,32 +54,24 @@ class ContinuousSuiteBase extends StreamTest {
   // A continuous trigger that will only fire the initial time for the duration of a test.
   // This allows clean testing with manual epoch advancement.
   protected val longContinuousTrigger = Trigger.Continuous("1 hour")
+
+  override protected val defaultTrigger = Trigger.Continuous(100)
+  override protected val defaultUseV2Sink = true
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: Why are these changes needed?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala [hunk 18]
```diff
@@ -53,32 +54,24 @@ class ContinuousSuiteBase extends StreamTest {
...
 
-  test("basic rate source") {
-    val df = spark.readStream
-      .format("rate")
-      .option("numPartitions", "5")
-      .option("rowsPerSecond", "5")
-      .load()
-      .select('value)
+  test("basic") {
+    val input = new ContinuousMemoryStream[Int](0, spark.sqlContext)
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: ContinuousMemoryStream can have no-arg apply, similar to MemoryStream. Must easier to use. 
Having the user specify the id is a recipe of disaster when the user gives the same number to multiple.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ContinuousMemoryStream.scala [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+ *  * ContinuousMemoryStream maintains a list of records for each partition. addData() will
+ *    distribute records evenly-ish across partitions.
+ *  * RecordEndpoint is set up as an endpoint for executor-side
+ *    ContinuousMemoryStreamDataReader instances to poll. It returns the record at the specified
+ *    offset within the list, or null if that offset doesn't yet have a record.
+ */
+class ContinuousMemoryStream[A : Encoder](id: Int, sqlContext: SQLContext)
+  extends MemoryStreamBase[A](sqlContext) with ContinuousReader with ContinuousReadSupport {
+  private implicit val formats = Serialization.formats(NoTypeHints)
+  val NUM_PARTITIONS = 2
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ContinuousMemoryStream.scala]: does this have to be public?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ContinuousMemoryStream.scala [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+    ContinuousMemoryStreamOffset(
+      offsets.map {
+        case ContinuousMemoryStreamPartitionOffset(part, num) => (part, num)
+      }.toMap
+    )
+  }
+
+  override def createDataReaderFactories(): ju.List[DataReaderFactory[Row]] = {
+    synchronized {
+      val endpointName = s"ContinuousMemoryStreamRecordReceiver-$id"
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ContinuousMemoryStream.scala]: id may not be unique. may be better to attach an UUID to always create a unique name and avoid any confusion in the future.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ContinuousMemoryStream.scala [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+  // ContinuousReadSupport implementation
+  // This is necessary because of how StreamTest finds the source for AddDataMemory steps.
+  def createContinuousReader(
+      schema: Optional[StructType],
+      checkpointLocation: String,
+      options: DataSourceOptions): ContinuousReader = {
+    this
+  }
+
+  override def reset(): Unit = synchronized {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ContinuousMemoryStream.scala]: Have you tested reusing this actually? does it work?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ContinuousMemoryStream.scala (hunk 20 above)
Comment (tdas) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ContinuousMemoryStream.scala]: Also, ContinuousMemoryStreamRecordReceiver -> ContinuousMemoryStreamRecordEndpoint

**Review by tdas — COMMENTED** [review]
Few more nits.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala (hunk 17 above)
Comment (jose-torres) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: to make testStream do continuous processing without explicitly specifying trigger and the flag

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ContinuousMemoryStream.scala (hunk 21 above)
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ContinuousMemoryStream.scala]: No.

I think we ought to just remove this method in a separate PR; it's only used in one place throughout the entire codebase and doesn't seem particularly useful. For now I've just removed it from ContinuousMemoryStream.

**Comment by tdas** [comment]
LGTM. Merging to master.
